Documentation/contributors/design-proposals/subscription-status.md

```diff
@@ -0,0 +1,188 @@
+    // ...
+)
+
+type SubscriptionStatus struct {
+    // ObservedGeneration is the generation observed by the Subscription controller.
+    // +optional
+    ObservedGeneration int64 `json:"observedGeneration,omitempty"`
+
+    // CurrentCSV is the CSV the Subscription is progressing to.
+    // +optional
+    CurrentCSV   string                `json:"currentCSV,omitempty"`
+
+    // InstalledCSV is the CSV currently installed by the Subscription.
+    // +optional
+    InstalledCSV string                `json:"installedCSV,omitempty"`
+
+    // Install is a reference to the latest InstallPlan generated for the Subscription.
+    // DEPRECATED
+    // +optional
+    Install      *InstallPlanReference `json:"installplan,omitempty"`
+
+    // State represents the current state of the Subscription
+    // +optional
+    State       SubscriptionState `json:"state,omitempty"`
+
+    // Reason is the reason the Subscription was transitioned to its current state.
+    // +optional
+    Reason      ConditionReason   `json:"reason,omitempty"`
+
+    // InstallPlanRef is a reference to the latest InstallPlan that contains the Subscription's current CSV.
+    // +optional
+    InstallPlanRef *corev1.ObjectReference
+
+
+    // UpToDate is true when the latest CSV for the Subscription's package and channel is installed and running; false otherwise.
+    //
+    // This field is not a status SubscriptionCondition because it "represents a well-known state that applies to all instances of a kind"
+    // (see https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api-conventions.md#typical-status-properties).
+    // In this case, all Subscriptions are either up to date or not up to date.
+    UpToDate bool `json:"UpToDate"`
+
+    // LastUpdated represents the last time that the Subscription status was updated
+    LastUpdated metav1.Time       `json:"lastUpdated"`
+
+    // Conditions is a list of the latest available observations about a Subscription's current state.
+    Conditions []SubscriptionCondition
+}
+
+// SubscriptionConditionType indicates an explicit state condition about a Subscription in "abnormal-true"
+// polarity form (see https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api-conventions.md#typical-status-properties).
+type SusbcriptionConditionType string
+
+const (
+    // SubscriptionFailed indicates the Subscription has failed progressing the operator installation to the latest version.
+    SubscriptionFailed SubscriptionConditionType = "Failed"
+
+    // SubscriptionProgressing indicates the Subscription is attempting to progress the cluster to the desired state.
+    SubscriptionProgressing SubscriptionConditionType = "Progressing"
+
+    // SubscriptionCatalogSourcesUnhealthy indicates that some or all of the CatalogSources to be used in resolution are unhealthy.
+    SubscriptionCatalogSourcesUnhealthy SubscriptionConditionType = "CatalogSourcesUnhealthy"
```

Comment:
Should some of these `SubscriptionConditionType`s be reduced to "XResourceIssue" with the specific details being in the `SubscriptionCondition.Reason` field? From the examples listed here it seems like the answer is no, but the API convention doc defines `Reason` scope as being used for kubectl output and doesn't use a wrapper type for the string.

Reply:
Should we make a rule of thumb for this?

How about: if we use the state to make decisions on actions to perform, then it's a condition, otherwise it's just a reason?

i.e. if we ever find ourselves doing `if x.Condition.Status = Y && x.Condition.Reason = Z`, then the Reason should be it's own Condition?

Reply:
By keeping the number of condition "types" small and generic while having many "reasons", we improve future backwards-compatibility by not relying on implementation details (that are still changing). The "type" should be a positive adjective or present progressive verb (`status: 'False'` would indicate negative condition). It would be easier to check for only "false" conditions and act on those according to the "reason".

See `status.conditions` for a failed pod:
```yaml
conditions:
- type: Ready
  status: 'False'
  reason: ContainersNotReady
  message: 'containers with unready status: [hello-openshift]'
  lastProbeTime: '2019-03-11T16:29:09Z'
  lastTransitionTime: '2019-03-11T16:29:09Z'
```

So a smaller set of condition types could contain all our rich reasons:
```go
const (
  // Possible reasons: "CatalogSourceInvalid", "PackageChannelInvalid", "InstallPlanFailed", etc
  SubscriptionConditionResolved  SubscriptionConditionType = "Resolved"
  // Possible reasons: "RequiresApproval", "AtLatest", "InstalledCSVReplacementAvailable", etc
  SubscriptionConditionProgressing  SubscriptionConditionType = "Progressing"
)
```

Reply:
@ecordell your rule of thumb sounds appealing - I like having a single, actionable decision point.
 
@alecmerdler I disagree with your definition for "types". I think they should be reserved for abnormal states rather than valid states, since valid should be the level/norm (as state approaches declared). Taking action also only makes sense when there exists an abnormal state, ie. progressing, unhealthy catalogsources, pods not ready, etc... Keep in mind, status has three types, `True`, `False, and `Unknown`, where the last of these indicates we need to check for the abnormal status. See [api-conventions docs](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api-conventions.md#typical-status-properties) where they talk about "abnormal-true" polarity. The same "check for 'False'" logic applies. 

In the future, as we look into statecharts, I'm thinking that each reasonable set of `Condition`s should get their own concurrent state machine.

I'm also not sure moving the details of what went wrong into `Reason` decouples us from implementation details. If `Reason` determines action, then we still need a state-machine that acts on a combination of "type" and `Reason` - we just split the predicate for state in two.